Split orphan handling from net_processing into txorphanage

ajtowns opened the pull request:
Splits orphan handling into its own module and reduces global usage.

ajtowns committed:
move-only: Add txorphanage module

This module captures orphan tracking code for tx relay.

Can be reviewed with --color-moved=dimmed-zebra

ajtowns committed:
txorphanage: Extract AddChildrenToWorkSet

Extract some common code into AddChildrenToWorkSet function.

(It's a hard knock life)

ajtowns committed:
txorphanage: Extract HaveOrphanTx

Extract some common code into HaveOrphanTx function.

ajtowns committed:
txorphanage: Extract GetOrphanTx

Extract orphan lookup code into GetOrphanTx function.

ajtowns committed:
txorphanage: Extract OrphanageAddTx

Extract code from AddOrphanTx into OrphanageAddTx.

ajtowns committed:
txorphanage: Extract EraseOrphansForBlock

Extract code that erases orphans when a new block is found into
EraseOrphansForBlock.

ajtowns committed:
txorphanage: Move functions and data into class

Collects all the orphan handling globals into a single member var in
net_processing, and ensures access is encapuslated into the interface
functions. Also adds doxygen comments for methods.

ajtowns committed:
net_processing: move AddToCompactExtraTransactions into PeerManagerImpl

Allows making vExtraTxnForCompact and vExtraTxnForCompactIt member vars
instead of globals.

ajtowns committed:
txorphanage: Add lock annotations

EraseOrphansFor was called both with and without g_cs_orphans held,
correct that so that it's always called with it already held.

LimitOrphanTxSize was always called with g_cs_orphans held, so
add annotations and don't lock it a second time.

ajtowns committed:
denialofservices_tests: check txorphanage's AddTx

Rather than checking net_processing's internal implementation of
AddOrphanTx, test txorphanage's exported AddTx interface. Note that
this means AddToCompactExtraTransactions is no longer tested here.

ajtowns commented:
Rebased for conflict with #21062

jnewbery commented:
Concept ACK. Will review once this is rebased and out of draft status.

ajtowns committed:
scripted-diff: Update txorphanage naming convention

-BEGIN VERIFY SCRIPT-
sed -i 's/mapOrphanTransactionsByPrev/m_outpoint_to_orphan_it/g' src/txorphanage.h src/txorphanage.cpp
sed -i 's/mapOrphanTransactions/m_orphans/g' src/txorphanage.h src/txorphanage.cpp src/net_processing.cpp src/test/denialofservice_tests.cpp
sed -i 's/g_orphan_list/m_orphan_list/g' src/txorphanage.h src/txorphanage.cpp
sed -i 's/g_orphans_by_wtxid/m_wtxid_to_orphan_it/g' src/txorphanage.h src/txorphanage.cpp
sed -i 's/nMaxOrphans/max_orphans/g' src/txorphanage.h src/txorphanage.cpp
sed -i 's/COrphanTx/OrphanTx/g' src/txorphanage.h src/txorphanage.cpp src/test/denialofservice_tests.cpp
-END VERIFY SCRIPT-

ajtowns committed:
txorphanage: Pass uint256 by reference instead of value

ajtowns commented:
Rebased, tweaked the orphan code a little, and cleaned up the results some. Should be ready for review.

amitiuttarwar commented:
concept ACK, starting to review. 

two thoughts for now-
1. I don't quite understand the context of the first commit. There are other commits that are essentially move-only, so wondering how you grouped together the changes in that commit. 
2. Are you open to some feedback around improvements of code you moved around? eg. using chrono instead of ints? or are you trying to keep the diff as minimal as possible? I personally think it'd be a good time for stuff that would be small, clear improvements, but won't leave those sorts of review comments if you're not interested :)

ajtowns commented:
@amitiuttarwar 1) The first commit just sets up the new files and moves whole functions / globals into it. Later commits are pulling code out into their own functions, restructuring functions into classes, etc. 2) hmm, I was ignoring chrono stuff since #21015 exists, but I guess it didn't switch COrphanTx to chrono. Not sure it makes sense to add more things to this PR versus adding to the #20758 backlog.

jnewbery reviewed (commented):
utACK 649cad4fc8fcc107a02a5e448cf9495d38e18add

Mostly style comments, with a couple of requests for minor changes to the interface that I think would improve the encapsulation.

ajtowns commented on src/test/denialofservice_tests.cpp:
```diff
@@ -295,15 +284,23 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
 }
 
-static CTransactionRef RandomOrphan()
+class TxOrphanageTest : public TxOrphanage
```
I think that would be a good idea for a later PR (particularly one that added more test coverage)

ajtowns commented on src/net_processing.cpp:
Not sure about the `const` part -- it might make sense for the limit to change depending on how much pressure there is in the mempool at large. Agree otherwise.

amitiuttarwar replied:
+1 came here to make a very similar suggestion :)

ajtowns committed:
net_processing: drop AddOrphanTx

All the interesting functionality of AddOrphanTx is already in other
functions, so call those functions directly in the one place that
AddOrphanTx was used.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -1133,126 +1102,17 @@ static void AddToCompactExtraTransactions(const CTransactionRef& tx) EXCLUSIVE_L
 }
 
-bool AddOrphanTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans)
+bool PeerManagerImpl::AddOrphanTx(const CTransactionRef& tx, NodeId peer)
```
Dang, why didn't I think of that.

ajtowns commented on src/txorphanage.h:
```diff
@@ -0,0 +1,81 @@
+    bool HaveTx(const GenTxid& gtxid) EXCLUSIVE_LOCKS_REQUIRED(!g_cs_orphans);
+
+    /** Get the details of an orphan transaction (returns nullptr if not found) */
+    OrphanTx* GetTx(const uint256& txid) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
```
Done. First instance of structure binding in the codebase?

ajtowns commented on src/txorphanage.h:
```diff
@@ -0,0 +1,81 @@
+    /** Erase all orphans announced by a peer (eg, after that peer disconnects) */
+    void EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+
+    /** Erase all orphans included in / invalidated by a new block */
+    void EraseForBlock(const CBlock& block) EXCLUSIVE_LOCKS_REQUIRED(!g_cs_orphans);
```
Definite nack on the comment change -- the function erases orphans that are double-spent by txs in the block, not just orphans that are included in the block directly.

I'm not against having vtx dereferenced in the caller, but the logic seems specific to a block arriving: if it were just an in-mempool tx double-spending an orphan tx, it might be that the orphan is actually a successful RBF and should be accepted once we know all its parents. So I think having it be the actual block as the argument makes sense; and since it's a lightweight header that's transitively included anyway, I've kept it as-is (but included the header explicitly).

jnewbery replied:
Ah, good point about erasing orphans that are conflicted by the block. I agree that it's fine to keep this as it is.

Thanks for including primitives/block.h

ajtowns commented:
Added most of jnewbery's suggestions, and also added some missing `const` annotations on `TxOrphanage` methods.

jnewbery reviewed (commented):
utACK eeeafb324ef6057f40b5c5fdd8464110e809b0f7

Two tiny style suggestions inline.

The last commit (_net_processing: move AddToCompactExtraTransactions into PeerManagerImpl_) doesn't really need to be in this PR. No harm in having it here, but could equally wait until the next PR (which could also change `AddToCompactExtraTransactions()`, `vExtraTxnForCompact` and `vExtraTxnForCompactIt` to be guarded by cs_main).

glozow reviewed (commented):
utACK eeeafb324e

I know this is mostly a move-only PR, but I kind of treated it as "a tour of orphan handling" so I have a bunch of questions, heh. Thanks for adding comments too!

amitiuttarwar commented on src/txorphanage.h:
```diff
@@ -13,30 +13,48 @@
- *  -maxorphantx/DEFAULT_MAX_ORPHAN_TRANSACTIONS */
-extern std::map<uint256, COrphanTx> mapOrphanTransactions GUARDED_BY(g_cs_orphans);
+    /** Erase all orphans announced by a peer (eg, after that peer disconnects) */
+    void EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
```
Is there a reason `peer` here and `AddTx`, `EraseForPeer` are non-const? I think they can be but since most things in this header are constified & it's consistently the peer NodeId that's not, I'm wondering if I'm missing something.

Also, I like the const member functions, that's awesome for communicating/enforcing the intention of the functions.

ajtowns replied:
I don't usually add `const` to arguments that are passed by vaue rather than reference -- it makes no difference to the caller; and not much difference to the implementation.

amitiuttarwar replied:
ah, right. thanks

amitiuttarwar commented on src/txorphanage.h:
```diff
@@ -0,0 +1,80 @@
+
+    /** Index from wtxid into the m_orphans to lookup orphan
+     *  transactions using their witness ids. */
+    std::map<uint256, OrphanMap::iterator> m_wtxid_to_orphan_it GUARDED_BY(g_cs_orphans);
```
hmm, it seems like a tradeoff between memory & runtime complexity. the `HaveTx` lookup is currently logarithmic, but using a `find_if` would be linear. I'm guessing opting for the additional data structure was chosen to minimize the runtime impact on the program since we call this function so frequently (at least.. for every txid in inv messages, before we send a getdata, when processing a tx)

amitiuttarwar reviewed (commented):
code review ACK eeeafb324e 

the orphanage is v. beautiful! my least favorite thing about it is the two-line if statements without parenthesis (they were copied over in move-only commits), but I'll survive :)

sipa commented on src/net_processing.cpp:
```diff
@@ -2104,10 +2104,9 @@ void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
 
-        auto orphan_it = mapOrphanTransactions.find(orphanHash);
-        if (orphan_it == mapOrphanTransactions.end()) continue;
+        const auto [porphanTx, from_peer] = GetOrphanTx(orphanHash);
```
Woah! I forgot we got that with C++17.

ajtowns commented on src/txorphanage.h:
```diff
@@ -0,0 +1,81 @@
+    struct OrphanTx {
+        CTransactionRef tx;
+        NodeId fromPeer;
+        int64_t nTimeExpire;
```
The `CTransactionRef tx` could be const as well; the only entry that changes after insertion is `list_pos`. Not sure there's much benefit to adding the const annotations for that though.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -1003,7 +1003,7 @@ void PeerManagerImpl::FinalizeNode(const CNode& node, bool& fUpdateConnectionTim
         mapBlocksInFlight.erase(entry.hash);
     }
-    EraseOrphansFor(nodeid);
+    WITH_LOCK(g_cs_orphans, EraseOrphansFor(nodeid));
```
denialofservices_tests locks `g_cs_orphans` then calls `EraseOrphansFor`, so it would have required more changes to have `EraseOrphansFor` do the locking.

I don't think `FinalizeNode` is called from anywhere that holds `g_cs_orphans`? Am I missing somewhere?

jnewbery replied:
Here: https://github.com/bitcoin/bitcoin/blob/b9f41df1ead4b6a83a51fc41966b111c8459c313/src/init.cpp#L213-L214 ?

ajtowns replied:
Yes, that's it!

StopNodes() calls DeleteNode() calls FinalizeNode() calls EraseForPeer(); with EraseForPeer needing `g_cs_orphans`. StopNodes locks `cs_vNodes` before calling DeleteNode, so `g_cs_orphans` has to be locked before calling CConnman::StopNodes but only if it's linked to a PeerManagerImpl.

StopNodes is also called from `~CConnman` so that path will still get a lock order violation error, I think.

DeleteNode is also called from DisconnectNodes which is called from ThreadSocketHandler; but there it's called after `cs_vNodes` is released, so lock ordering is not an issue, but `g_cs_orphans` still has to get locked prior to doing the work in EraseForPeer, so responsibility can't be moved all the way back up the chain. 

So I don't think there's a simple way to make it non-recursive for this PR; for a future PR, I think the aim is to make `g_cs_orphans` become `private: TxOrphanage::m_mutex` and not have it be the last thing locked / first thing released so that the ordering constraint goes away, and then it should trivially be non-recursive.

jnewbery replied:
> for a future PR, I think the aim is to make g_cs_orphans become private: TxOrphanage::m_mutex and not have it be the last thing locked / first thing released so that the ordering constraint goes away, and then it should trivially be non-recursive.

Concept ACK :)

ajtowns replied:
https://github.com/ajtowns/bitcoin/commits/202102-orphanworkset has work-in-progress that turns g_cs_orphans a protected m_mutex, doesn't quite turn orphans back into the responsibility of who sent them rather than who sent their final parent though.

ajtowns committed:
txorphanage: comment improvements

ajtowns commented:
Added a commit to update comments with suggestions.

@jnewbery yeah, other commit could have been dropped -- it was only there to move `AddOrphan` into `PeerManagerImpl` so that `m_orphanage` could also be in `PeerManagerImpl`, but that stopped being an issue after doing away with net_processing's `AddOrphan` -- but should be pretty harmless to leave it there.

jnewbery commented:
utACK 5e50e2d1b9

amitiuttarwar commented:
utACK 5e50e2d1b95e7ca7709a9671ab21f1164b8d0cb8

glozow commented:
re ACK https://github.com/bitcoin/bitcoin/pull/21148/commits/5e50e2d1b95e7ca7709a9671ab21f1164b8d0cb8, comment updates

laanwj commented:
> the orphanage is v. beautiful! my least favorite thing about it is the two-line if statements without parenthesis (they were copied over in move-only commits), but I'll survive :)

Yes I like how this cleanly encapsulates the transaction orphan handling, exposing a single documented interface in a small header file.

It is clearly an improvement. The only thing I don't completely understand is why the `g_cs_orphans` lock is global and not on the orphanage itself, but maybe that's too much impact? I think lock refactoring could be a topic for a future PR, anyhow, no need to hold this up.

Code review ACK 5e50e2d1b95e7ca7709a9671ab21f1164b8d0cb8

jnewbery commented:
>  The only thing I don't completely understand is why the g_cs_orphans lock is global and not on the orphanage itself, but maybe that's too much impact?

`g_cs_orphans` is still used to guard some objects outside the orphange (`Peer.m_orphan_work_set` and `vExtraTxnForCompact`). The plan is for future PRs to remove those dependencies and make the orphan mutex a member of the orphanage.

laanwj commented:
@jnewbery Good to know, thanks!

ariard commented on src/txorphanage.cpp:
```diff
@@ -0,0 +1,202 @@
+    return {it->second.tx, it->second.fromPeer};
+}
+
+void TxOrphanage::EraseForBlock(const CBlock& block)
```
If we had a `std::map<uint256, OrphanMap::iterator> m_parentid_to_orphan_it` index we could update each peer's `orphan_work_set` in function of the parent received through this block. 

Otherwise, I think you might have stalling valid orphans never re-processed. Any ulterior parent announcement should be bounce off by `m_recent_confirmed_transactions`. Though rare enough to not be worthy of the complexity...

ariard commented on src/txorphanage.h:
```diff
@@ -0,0 +1,85 @@
+    /** Check if we already have an orphan transaction (by txid or wtxid) */
+    bool HaveTx(const GenTxid& gtxid) const EXCLUSIVE_LOCKS_REQUIRED(!g_cs_orphans);
+
+    /** Get an orphan transaction and its orginating peer
```
nit: originating

ariard reviewed (commented):
Post-Merge Code Review ACK 5e50e2d

MarcoFalke commented:
pm ACK

ajtowns commented on src/txorphanage.cpp:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+bool TxOrphanage::HaveTx(const GenTxid& gtxid) const
```
I think what you're saying is: Bob has tx `P` (parent) and a child of that tx `C`. Mallory knows both and malleates the witness of `C` so that it creates a new invalid tx `X` that has the same txid as `C` but obviously a different wtxid. Mallory relays `X` to Alice before Alice hears about `P` or `C`. Alice adds `X` to the orphan pool and requests `P` from Mallory by wtxid.

I think the expected behaviour is then: Carol announces both `P` and `C` to Alice; if wtxid relay is active, both are announced via wtxid so don't conflict with `X`, and both are requested. If the request for `P` from Mallory is still active, `C` will be requested first, and when we receive it, we'll see there's already an entry for that txid in the orphan pool and ignore it. I'm not sure if we'll re-request from other honest peers at that point in normal usage; though I think this is a case where #21061 would eventually save the day.

ariard replied:
> I'm not sure if we'll re-request from other honest peers at that point in normal usage

Once the transaction is rejected from orphanage in reason of an already present orphan, we forget its txid/wtxid from tx-requester. So effectively, we shouldn't re-request it from other onest peers at that stage. Ultimately, parent `P` withhold by Mallory should expire, Alice will fetch a valid parent `P` from another peer. When Bob rebroadcasts `C`, Mallory isn't able to interfere anymore as malleated `X` will be rejected directly, without being logged in orphan pool.

Yes I agree #21061 should make things better.
